crates/red_knot_python_semantic/src/types/mro.rs

```diff
@@ -0,0 +1,641 @@
+            Self::Class(class) => class.own_class_member(db, member),
+        }
+    }
+
+    pub(super) fn display(self, db: &'db dyn Db) -> Cow<'static, str> {
+        match self {
+            Self::Any => Cow::Borrowed("Any"),
+            Self::Todo => Cow::Borrowed("Todo"),
+            Self::Unknown => Cow::Borrowed("Unknown"),
+            Self::Class(class) => Cow::Owned(format!("<class '{}'>", class.name(db))),
+        }
+    }
+
+    /// Return the various [`MroPossibilities`] for this base.
+    fn mro_possibilities(self, db: &'db dyn Db) -> Cow<MroPossibilities<'db>> {
+        match self {
+            ClassBase::Class(class) => Cow::Borrowed(class.mro_possibilities(db)),
+            ClassBase::Any | ClassBase::Todo | ClassBase::Unknown => {
+                let object = ClassBase::builtins_object(db);
+                Cow::Owned(MroPossibilities::single([self, object]))
+            }
+        }
+    }
+}
+
+impl<'db> From<Type<'db>> for ClassBase<'db> {
+    fn from(value: Type<'db>) -> Self {
+        match value {
+            Type::Any => ClassBase::Any,
+            Type::Todo => ClassBase::Todo,
+            Type::Unknown => ClassBase::Unknown,
+            Type::Class(class) => ClassBase::Class(class),
+            // TODO support `__mro_entries__`?? --Alex
+            Type::Instance(_) => ClassBase::Todo,
+            // TODO: ??
+            Type::Intersection(_) => ClassBase::Todo,
+            Type::Union(_) => {
+                panic!(
+                    "Should never call `ClassBase::from` on a `Type::Union` variant; \
+                    unions have custom handling throughout"
+                )
+            }
+            // These are all errors:
+            Type::Unbound
+            | Type::BooleanLiteral(_)
+            | Type::BytesLiteral(_)
+            | Type::Function(_)
+            | Type::IntLiteral(_)
+            | Type::LiteralString
+            | Type::Module(_)
+            | Type::Never
+            | Type::None
+            | Type::StringLiteral(_)
+            | Type::Tuple(_) => ClassBase::Unknown,
+        }
+    }
+}
+
+impl<'db> From<&Type<'db>> for ClassBase<'db> {
+    fn from(value: &Type<'db>) -> Self {
+        Self::from(*value)
+    }
+}
+
+impl<'db> From<ClassBase<'db>> for Type<'db> {
+    fn from(value: ClassBase<'db>) -> Self {
+        match value {
+            ClassBase::Class(class) => Type::Class(class),
+            ClassBase::Any => Type::Any,
+            ClassBase::Todo => Type::Todo,
+            ClassBase::Unknown => Type::Unknown,
+        }
+    }
+}
+
+impl<'db> From<&ClassBase<'db>> for Type<'db> {
+    fn from(value: &ClassBase<'db>) -> Self {
+        Self::from(*value)
+    }
+}
+
+/// Represents a class's bases list.
+///
+/// At runtime, this list of classes can be found
+/// in the `__bases__` attribute on a class.
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Default)]
+pub(super) struct BasesList<'db>(Box<[ClassBase<'db>]>);
```

Comment:
What's the meaning of an empty base list. Is it if there's no explicit base?

Reply:
The `BasesList` represents all bases the class has (both implicit and explicit). For a class that implicitly inherits from `object`, e.g. `class Foo: pass`, the bases of `Foo` will be `[object]`. Only one class has an empty bases list in Python: `object` itself:

```pycon
>>> object.__bases__
()
>>> class Foo: pass
... 
>>> Foo.__bases__
(<class 'object'>,)
```